osbuild/util/pathfd.py

```diff
@@ -0,0 +1,716 @@
+        cannot directly open a path-descriptor, but have to redirect via
+        `/proc/self/fd/<fd>`. Hence, this is what this implementation does.
+        This is, however, not strictly equivalent to opening the
+        path-descriptor directly. Therefore, some features are blocked to avoid
+        exposing broken behavior:
+
+          * You cannot use `O_NOFOLLOW` with this, as it would open the symlink
+            in `/proc` rather than the object the descriptor points to (and
+            because you cannot open symlinks, it would fail). Hence, passing
+            `O_NOFOLLOW` will raise an exception.
+            If you want this behavior, you must check via `is_symlink()`
+            manually, and then call this without `O_NOFOLLOW`.
+
+          * This relies on `/proc` being a proper `procfs` file-system. This
+            is the case for all known linux systems. If you use non-standard
+            setups, you must be aware that this call will fail, unless `/proc`
+            is properly setup (or worse if you re-use `/proc` for other
+            things).
+
+        This function always forces `O_CLOEXEC`. You must clear it afterwards,
+        if that is what you want.
+
+        Parameters
+        ----------
+        flags : int
+            The open-flags to pass to `os.open()`.
+        mode : int, optional
+            The file creation flags to use with `os.open()`. Defaults to
+            `0o777`, but is subject to the current umask.
+
+        Returns
+        -------
+        int
+            The syscall result of `open(2)` is returned. This is usually a
+            file-descriptor.
+
+        Raises
+        ------
+        OSError
+            See the python standard library for possible errors when opening
+            paths (e.g., `os.open()`).
+        """
+
+        #
+        # Preferably, we would want to use `openat(self, "", ... O_EMPTYPATH)`,
+        # but sadly there is no `AT_EMPTY_PATH` equivalent for `openat(2)`.
+        # Instead, we must go the route via `/proc`.
+        # Note that for questionable security reasons the `O_EMPTYPATH` flag is
+        # unlikely to appear at all (also see `AT_EMPTY_PATH` in `linkat(2)`
+        # for a discussion).
+        #
+        # We simply block the `O_NOFOLLOW` flag to make sure callers will not
+        # accidentally end up opening a file in `/proc`. If you want
+        # `O_NOFOLLOW` with `open_self()`, you are left with checking through
+        # `is_symlink()` before calling `open_self()` without `O_NOFOLLOW`.
+        # Note that this is safe as long as you call it on the same `DirFd`.
+        #
+
+        assert self.is_open()
+        assert not (flags & _os.O_NOFOLLOW)
+
+        flags |= _os.O_CLOEXEC
+        path = _os.path.join("/proc/self/fd/", str(int(self)))
+
+        return _os.open(path, flags, mode)
+
+    def descend(self, path, *, follow_symlink = False):
```

Comment:
Yeah, I intentionally left this out for now.

First of all, you can achieve the same via `p.descend("go").descend("much").descend("deeper")`, which descends down a file-system tree `go/much/deeper`. Of course, this is a lot more expensive, it is no longer atomic (on linux, even passing "go/much/deeper" directly is not atomic either, as it uses RCU-traversal in the kernel), and it is very hard to read. However, in some cases it might be what you want.

Secondly, you can always do `p.descend(os.path.join("go", "much", "deeper"))`, if you want this to join file-system paths.

However, I do see the similarity to `os.path.join(2)`. There are some differences. `join()` is a global while `descend()` is a method-call. `join()` takes the same type as it returns, while `descend()` takes a path and turns it into a pathfd. Maybe that affects how we end up using the function, and thus we might not really end up joining paths on the call-site anyway?
I liked being completely agnostic of path-handling, not having to know about slashes, dot and dotdot, nor any other special behavior of paths, but just letting the kernel deal with it. Nevertheless, I am definitely open to extending this if we end up doing joins on the call-site and it just feels natural to add it.

Long story short: It was mostly a "I don't want to deal with paths" decision, rather than a blocker against doing this.